stk/calculators/optimization/optimizers.py

```diff
@@ -648,102 +650,501 @@ def optimize(self, mol, conformer=-1):
+
+        if self.solvent is not None:
+            solvent = f'--gbsa {self.solvent} {self.solvent_grid}'
+        else:
+            solvent = ''
+
+        cmd = (
+            f'{memory} {self.xtb_path} {xyz} --gfn {self.gfn_version} '
+            f'{optimization} --parallel {self.num_cores} '
+            f'--etemp {self.electronic_temperature} '
+            f'{solvent} --chrg {self.charge} '
+            f'--uhf {self.num_unpaired_electrons}'
+        )
+
+        with open(out_file, 'w') as f:
+            # Note that sp.call will hold the program until completion
+            # of the calculation.
+            sp.call(
+                cmd,
+                stdin=sp.PIPE,
+                stdout=f,
+                stderr=sp.PIPE,
+                # Uses the shell if unlimited_memory is True to be run
+                # multiple commpands in one subprocess.
+                shell=self.unlimited_memory
+            )
+
+    def _run_optimizations(self, mol, conformer):
+        """
+        Run loop of optimizations of the molecule `mol` using xTB.
+
+        Parameters
+        ----------
+        mol : :class:`.Molecule`
+            The molecule to be optimized.
+
+        conformer : :class:`int`
+            The conformer to use.
+
+        Returns
+        -------
+        None : :class:`NoneType`
+
+        """
+        for run in range(self.max_runs):
+            xyz = f'input_structure_{run+1}.xyz'
+            out_file = f'optimization_{run+1}.output'
+            mol.write(xyz, conformer=conformer)
+            self._run_xtb(xyz=xyz, out_file=out_file)
+            # Check if the optimization is complete.
+            coord_file = 'xtbhess.coord'
+            coord_exists = os.path.exists(coord_file)
+            output_xyz = 'xtbopt.xyz'
+            opt_incomplete = self._incomplete(out_file)
+            if opt_incomplete and coord_exists:
+                # The calculation is incomplete.
+                # Update mol from xtbhess.coord and continue.
+                mol.update_from_turbomole(
+                    path=coord_file,
+                    conformer=conformer
+                )
+            else:
+                # Update mol from xtbopt.xyz.
+                mol.update_from_xyz(
+                    path=output_xyz,
+                    conformer=conformer
+                )
+                if opt_incomplete:
+                    # If the negative frequencies are small, then GFN
+                    # may not produce the restart file. If that is the
+                    # case, exit optimization loop and warn.
+                    self.incomplete.add((mol, conformer))
+                    logging.warning(
+                        'Small negative frequencies present in '
+                        f'{mol.name} conformer {conformer}.'
+                    )
+                    return False
+                else:
+                    # Calculation is complete.
+                    return True
```

Comment:
I think the nesting here can be made a little bit clearer

```python

            if opt_incomplete:
                if coord_exists:
                    # The calculation is incomplete.
                    # Update mol from xtbhess.coord and continue.
                    mol.update_from_turbomole(
                        path=coord_file,
                        conformer=conformer
                    )
                else:
                    # Update mol from xtbopt.xyz.
                    mol.update_from_xyz(
                        path=output_xyz,
                        conformer=conformer
                    )
                    # If the negative frequencies are small, then GFN
                    # may not produce the restart file. If that is the
                    # case, exit optimization loop and warn.
                    self.incomplete.add((mol, conformer))
                    logging.warning(
                        'Small negative frequencies present in '
                        f'{mol.name} conformer {conformer}.'
                    )
                    return False
            else:
                break
        return True

```

Note the addition of return True at the end of the function. and the insertion of break in the loop.

Reply:
this loop does not update the mol from the xyz if opt_incomplete is False. That was the reason for the weird nesting.

Reply:
The way its written at the moment, it does update from xyz when opt_incomplete is False and coord_exists is True